drivers/net/ethernet/intel/idpf/eth/idpf_eth.c

```diff
@@ -0,0 +1,416 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/* Copyright (C) 2023 Intel Corporation */
+
+#include "idpf_eth.h"
+#include "idpf_netdev.h"
+
+/* Forward declaration */
+static void idpf_eth_device_post_init_task(struct work_struct *work);
+
+/**
+ * idpf_eth_statistics_task - Delayed task to get statistics over mailbox
+ * @work: work_struct handle to eth adapter data
+ */
+void idpf_eth_statistics_task(struct work_struct *work)
+{
+	struct idpf_eth_adapter *adapter;
+	struct idpf_vport *vport;
+
+	adapter = container_of(work, struct idpf_eth_adapter, stats_task.work);
+	vport = adapter->vport;
+	if (vport && !test_bit(IDPF_ETH_RESET_IN_PROG, adapter->flags))
+		idpf_send_get_stats_msg(vport);
+
+	queue_delayed_work(adapter->stats_wq, &adapter->stats_task,
+			   msecs_to_jiffies(10000));
+}
+
+/**
+ * idpf_eth_adapter_alloc - Allocate ethernet adapter struct
+ * @dev: Device struct
+ */
+static struct idpf_eth_adapter *idpf_eth_adapter_alloc(struct device *dev)
+{
+	struct idpf_eth_adapter *adapter;
+
+	adapter = kzalloc(sizeof(*adapter), GFP_KERNEL);
+	if (!adapter)
+		return NULL;
+
+	adapter->post_init_wq = alloc_workqueue("%s-%s-pinit", 0, 0,
+						dev_driver_string(dev),
+						dev_name(dev));
+	if (!adapter->post_init_wq) {
+		dev_err(dev, "Failed to allocate post init workqueue\n");
+		kfree(adapter);
+		return NULL;
+	}
+
+	adapter->stats_wq = alloc_workqueue("%s-%s-stats", 0, 0,
+					    dev_driver_string(dev),
+					    dev_name(dev));
+	if (!adapter->stats_wq) {
+		dev_err(dev, "Failed to allocate workqueue\n");
+		kfree(adapter);
+		goto err_stats_wq;
+	}
+
+	INIT_DELAYED_WORK(&adapter->post_init_task,
+			  idpf_eth_device_post_init_task);
+	INIT_DELAYED_WORK(&adapter->stats_task, idpf_eth_statistics_task);
+	mutex_init(&adapter->vport_ctrl_lock);
+
+	return adapter;
+
+err_stats_wq:
+	destroy_workqueue(adapter->post_init_wq);
+	kfree(adapter);
+	return NULL;
+}
+
+/**
+ * idpf_eth_device_post_init - Delayed ethernet post initialization
+ * @adapter: Ethernet private data struct
+ */
+static void idpf_eth_device_post_init(struct idpf_eth_adapter *adapter)
+{
+	struct idpf_vport_config *vport_config;
+
+	vport_config = adapter->vport_config;
+	/* Vport/Netdev is ready for initialization */
+	if (adapter->netdev &&
+	    !test_bit(IDPF_VPORT_REG_NETDEV, vport_config->flags)) {
+		register_netdev(adapter->netdev);
+		set_bit(IDPF_VPORT_REG_NETDEV, vport_config->flags);
+	}
+
+	/* Mark the ethernet driver initialize completion */
+	set_bit(IDPF_ETH_INIT_COMPLETE, adapter->flags);
+
+	/* Start the statistics task now */
+	queue_delayed_work(adapter->stats_wq, &adapter->stats_task, 0);
+}
+
+/**
+ * idpf_eth_device_post_init_task - Delayed ethernet post initialization task
+ * @work: work_struct handle to ethernet adapter data
+ */
+static void idpf_eth_device_post_init_task(struct work_struct *work)
```

Comment:
I don't see why this is a task. There is nothing blocking here. So why spawn a task ?

Reply:
Please see this flag "adapter->start_post_init" in this function. It is a blocking factor to re spawn.